Configuration Profiles documentation (#738)
awslabs/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/ConfigurationPresets.md

# Configuration Presets

> 

Reviewer: ```suggestion
## What is a Configuration Preset?

A Configuration Preset is a [configuration profile](./UsingTheJdbcDriver.md#configuration-profiles) that has already been set up by the AWS JDBC Driver team. Preset configuration profiles are optimized, profiled, verified and can be used right away. If the existing presets do not cover an exact use case, users can also create their own configuration profiles based on the built-in presets.
```

## Using of Configuration Presets

Configuration Preset name should be specified in [`wrapperProfileName`](#connection-plugin-manager-parameters) parameter.

```java
properties.setProperty("wrapperProfileName", "A2");
```

> User can create their own custom configuration profiles based on built-in configuration presets. 

Reviewer: ```suggestion
Users can create their own custom configuration profiles based on built-in configuration presets. 
```

Users can not delete built-in configuration presets.

```java
// Create a new configuration profile "myNewProfile" based on "A2" configuration preset
ConfigurationProfileBuilder.from("A2")
    .withName("myNewProfile")
    .withDialect(new CustomDatabaseDialect())
.buildAndSet();

properties.setProperty("wrapperProfileName", "myNewProfile");
```

## Existing Configuration Presets

Configuration Presets are optimized for 3 main user scenarios. They are:

> - No connection pool: presets family `A`, `B`, `C`
> - AWS JDBC Driver Internal connection pool: presets family `D`, `E`, `F`
> - External connection pool: presets family `G`, `H`, `I`

Reviewer: ```suggestion
- No connection pool preset family: `A`, `B`, `C`
- AWS JDBC Driver Internal connection pool preset family: `D`, `E`, `F`
- External connection pool preset family: `G`, `H`, `I`
```

> Some preset names may include a number, like `A0`, `A1`, `A2`, `D0`, `D1`, etc. Usually, the number represent a timing variations or sensitivity for the same preset. For example, `A0` is optimized for normal network outages sensitivity and normal response time, while `A1` is less sensitive. Please take into account that more aggressive presets tend to cause more false positive detections. For more details check [ConfigurationProfilePresetCodes.java](./../../wrapper/src/main/java/software/amazon/jdbc/profile/ConfigurationProfilePresetCodes.java) 

Reviewer: ```suggestion
Some preset names may include a number, like `A0`, `A1`, `A2`, `D0`, `D1`, etc. Usually, the number represent sensitivity or timing variations for the same preset. For example, `A0` is optimized for normal network outage sensitivity and normal response time, while `A1` is less sensitive. Please take into account that more aggressive presets tend to cause more false positive failure detections. More details can be found in this file: [ConfigurationProfilePresetCodes.java](./../../wrapper/src/main/java/software/amazon/jdbc/profile/ConfigurationProfilePresetCodes.java). 
```

```java
// Create a new configuration profile "myNewProfile" based on "A2" configuration preset
ConfigurationProfileBuilder.from("A2")
    .withName("myNewProfile")
    .withDialect(new CustomDatabaseDialect())
.buildAndSet();

properties.setProperty("wrapperProfileName", "myNewProfile");
```

## Existing Configuration Presets

Configuration Presets are optimized for 3 main user scenarios. They are:
- **No connection pool** preset family: `A`, `B`, `C`
- AWS JDBC Driver **Internal connection pool** preset family: `D`, `E`, `F`
- **External connection pool** preset family: `G`, `H`, `I`

Some preset names may include a number, like `A0`, `A1`, `A2`, `D0`, `D1`, etc. Usually, the number represent sensitivity or timing variations for the same preset. For example, `A0` is optimized for normal network outage sensitivity and normal response time, while `A1` is less sensitive. Please take into account that more aggressive presets tend to cause more false positive failure detections. More details can be found in this file: [ConfigurationProfilePresetCodes.java](./../../wrapper/src/main/java/software/amazon/jdbc/profile/ConfigurationProfilePresetCodes.java) 

> Choosing a right configuration preset for your application could be a challenging task. Many presets could potentially fit the needs of your application. Various user application requirements and goals are presented in the following table and organized to help you to identify most suitable presets.

Reviewer: ```suggestion
Choosing the right configuration preset for your application can be a challenging task. Many presets could potentially fit the needs of your application. Various user application requirements and goals are presented in the following table and organized to help you identify the most suitable presets for your application.
```
